SimplenoteTests/NoticeContorllerTests.swift

```diff
@@ -1,33 +1,142 @@
+    func testControllerPresent() {
+        let expectation = XCTestExpectation(description: "Did dismiss notice")
+        let timer = Timer.scheduledTimer(withTimeInterval: TimeInterval(0.5), repeats: false) { (_) in
+            expectation.fulfill()
+        }
+
+        let notice = Notice(message: "Message", action: nil)
+        controller.present(notice, withTimer: timer)
+
+        wait(for: [expectation], timeout: TimeInterval(3))
+
     }
 
-    func testExample() throws {
-        // This is an example of a functional test case.
-        // Use XCTAssert and related functions to verify your tests produce the correct results.
+    func testMakeNoticeView() {
+        let notice = Notice(message: "Message", action: nil)
+
+        controller.present(notice)
+        XCTAssertEqual(notice.message, presenter.noticeView?.message)
+        XCTAssertNil(presenter.noticeView?.handler)
     }
 
-    func testPerformanceExample() throws {
-        // This is an example of a performance test case.
-        self.measure {
-            // Put the code you want to measure the time of here.
+    func testMakeNoticeViewWithAction() {
+        let action = NoticeAction(title: "Title") {
+            print("Action")
         }
+        let notice = Notice(message: "Message ", action: action)
+
+        controller.present(notice)
+        XCTAssertEqual(notice.message, presenter.noticeView?.message)
+        XCTAssertEqual(notice.action?.title, presenter.noticeView?.actionTitle)
+        XCTAssertNotNil(presenter.noticeView?.handler)
+    }
+
+    func testMakeNoticeViewEmptyNotice() {
+        let action = NoticeAction(title: "") {
+            print("")
+        }
+        let notice = Notice(message: "", action: action)
+
+        controller.present(notice)
+        XCTAssertEqual(notice.message, presenter.noticeView?.message)
+        XCTAssertNotEqual(notice.action?.title, presenter.noticeView?.actionTitle)
+        XCTAssertNotNil(presenter.noticeView?.handler)
     }
 
+    func testDismiss() {
+        let noticeA = Notice(message: "Message A", action: nil)
+
+        controller.present(noticeA, withTimer: NoticeContorllerTests.noActionTimer)
+
+        controller.dismiss()
+
+        XCTAssertTrue(presenter.dismissed)
+        XCTAssertFalse(controller.isPresenting)
+    }
+
+    func testDismissContinuesToNextNotice() {
+        let noticeA = Notice(message: "Message A", action: nil)
+        let noticeB = Notice(message: "Message B", action: nil)
+        controller.present(noticeA, withTimer: NoticeContorllerTests.noActionTimer)
+        controller.present(noticeB, withTimer: NoticeContorllerTests.noActionTimer)
+
+        XCTAssertEqual(controller.pendingNotices, 1)
+
+        controller.dismiss()
+
+        XCTAssertTrue(presenter.dismissed)
+        XCTAssertTrue(controller.isPresenting)
+        XCTAssertEqual(controller.pendingNotices, 0)
+    }
```

Comment:
My understanding is that we want to test here that notices are presented one after the other. The best way to do that is to have a log of events in `MockNoticePresenter`. 
For example presenting 2 notices should result in the following log: `present(noticeA), dismiss(noticeA), present(noticeB), dismiss(noticeB)`. Later in the test we can compare the log with what we expect. 
I think this way the behaviour is more deterministic.

Reply:
Okay @eshurakov I have updated the tests as well.  I left a few questions for you in slack, so not all of these comments have been handled, but they are ready for another look.  Thanks!!!